vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,625 @@
+                }
+                descriptor = descriptors[0];
+
+                let len = desc_chain
+                    .memory()
+                    .write(event.to_bytes().as_slice(), descriptor.addr())
+                    .unwrap();
+
+                if vring.add_used(desc_chain.head_index(), len as u32).is_err() {
+                    println!("Couldn't write event data to the ring");
+                    return Err(VuInputError::UnexpectedWriteVringError);
+                }
+            } else {
+                // Now cannot get avaliable descriptor, which means the host cannot process
+                // event data in time and overrun happens in the backend. In this case,
+                // we simply drop the incomping input event and notify guest for handling
+                // events. At the end, it returns Ok(false) so can avoid exiting the thread loop.
+                self.ev_list.clear();
+
+                vring
+                    .signal_used_queue()
+                    .map_err(|_| VuInputError::SendNotificationFailed)?;
+
+                return Ok(false);
+            }
+        }
+
+        index = 0;
+        while index < last_sync_index {
+            self.ev_list.remove(0);
+            index += 1;
+        }
+
+        Ok(true)
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_queue(&mut self, vring: &VringRwLock) -> Result<bool> {
+        let evs = self.ev_dev.fetch_events().unwrap();
+
+        for ev in evs {
+            let ev_raw_data = VuInputEvent {
+                ev_type: ev.event_type().0,
+                code: ev.code(),
+                value: ev.value() as u32,
+            };
+            self.ev_list.push_back(ev_raw_data);
+        }
+
+        self.process_event(vring)?;
+
+        vring
+            .signal_used_queue()
+            .map_err(|_| VuInputError::SendNotificationFailed)?;
+
+        Ok(true)
+    }
+
+    pub fn read_event_config(&self) -> Result<VuInputConfig> {
+        let mut counter: u8 = 0;
+        let mut index: u8 = 0;
+        let mut cfg: [u8; VIRTIO_INPUT_CFG_SIZE] = [0; VIRTIO_INPUT_CFG_SIZE];
+
+        let func: unsafe fn(nix::libc::c_int, &mut [u8]) -> nix::Result<libc::c_int> =
+            match self.subsel {
+                EV_KEY => eviocgbit_key,
+                EV_ABS => eviocgbit_absolute,
+                EV_REL => eviocgbit_relative,
+                EV_MSC => eviocgbit_misc,
+                EV_SW => eviocgbit_switch,
+                _ => {
+                    return Err(VuInputError::HandleEventUnknownEvent);
+                }
+            };
+
+        // SAFETY: Safe as the file is a valid event device, the kernel will only
+        // update the correct amount of memory in func.
+        if unsafe { func(self.ev_dev.get_raw_fd(), &mut cfg) }.is_err() {
+            return Err(VuInputError::UnexpectedInputDeviceError);
+        }
+
+        for val in cfg.iter() {
```

Comment:
No need for the `.iter()` you can just do `for val in &cfg {}`

Reply:
Refined it.